src/Razor/Microsoft.AspNetCore.Razor.Language/src/HtmlConventions.cs

```diff
@@ -21,8 +22,19 @@ public static class HtmlConventions
                 RegexOptions.None,
                 TimeSpan.FromMilliseconds(500));
 
-        internal static IReadOnlyCollection<char> InvalidNonWhitespaceHtmlCharacters { get; } = new List<char>(
-            new[] { '@', '!', '<', '/', '?', '[', '>', ']', '=', '"', '\'', '*' });
+
+        internal static bool IsInvalidNonWhitespaceHtmlCharacters(char testChar)
+        {
+            foreach (var c in InvalidNonWhitespaceHtmlCharacters)
```

Comment:
Does this not allocate an enumerator? Wondering why not a for loop is all 😄

Reply:
The compiler optimizes a foreach for arrays - https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYZQAs6AsgBQCU+hxTAbgIYBO6H7rAnugF50AOwCmAdwDaAXXzoA5K3kAaBQCMVCgMbz0AXwDcjJkWMmA9OfQBVAM4BLYQHN0AMwD27Uay0ALdAA27u4ADmZMAGKeAKI+vtTcfLRGSCbE4cSWNg7Obp6BwWGpaehR7AnsPLzJGabFRHpmZuSlMXHUfhwyXJVJZgT1Jh5ecejUbJxa6I49VbS1DIMlOO7Ctu4BogB0AOrs9gAuoh01S8SNgxdMzRRlHb5dsonV/QvDYxPTgugADAZfAB5ZnwtgAZUTOA6+f72MBgeZnIgDZZMbCrdabXb7I4VKqSezSU7LK7nZB6IA===

Reply:
oh wow, so the compiler special cases arrays or something?

Reply:
> the compiler special cases arrays

It's not dictated by the language specifications, but as an optional optimization .NET's C# compiler (Roslyn) special cases (single dimensional) arrays, and strings so code equivalent to a for-loop is emitted.